mirage/awa_mirage.ml

```diff
@@ -107,32 +135,50 @@ module Make (F : Mirage_flow.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) =
       end
     | Error () -> match t.state with
       | `Error e -> Lwt.return (Error e)
-      | `Eof -> Lwt.return (Ok `Eof)
-      | `Active _ -> assert false
+      | `Closed | `Read_closed _ | `Active _ | `Write_closed _ -> Lwt.return (Ok `Eof)
 
   let close t =
-    (* TODO ssh session teardown (send some protocol messages) *)
-    FLOW.close t.flow >|= fun () ->
-    t.state <- `Eof
+    (match t.state with
+     | `Active ssh | `Read_closed ssh | `Write_closed ssh ->
+       let ssh, msg = Awa.Client.close ssh in
+       t.state <- inject_state ssh t.state;
+       writev_flow t (Option.to_list msg) >|= fun _ ->
+       t.state <- `Closed
+     | `Error _ | `Closed -> Lwt.return_unit) >>= fun () ->
+    FLOW.close t.flow
```

Comment:
If I follow correctly, if the state is `Write_closed`, we should not be able to write into the underlying flow. I'm right?

Reply:
I mean, if I look on this code:
https://github.com/mirage/awa-ssh/pull/70/files#diff-601db3741186b537bd27d174f1dbc9d54114a774e7f0a4ce889c60074daaddb7R53-R58

We see a transition to `Write_closed` and we should not write further bytes on the underlying bytes.

Reply:
If the state is `` `Write_closed `` then it means the *application* doesn't write anymore and the nexus sends a `Channel_eof` to the peer. So in that state the application can still read data and the nexus can also still send non-channel data messages. I think that's correct.

Now that you made me think about it I'm not sure about shutdown -- even if we send `Channel_eof` we might still want to send `Channel_eof` at a later point -- so we shouldn't ``shutdown`write`` the flow yet.

Reply:
So, the RFC is slightly unclear -- a EOF may be sent to indicate that no further data will be transmitted on that channel.

"A party MAY send SSH_MSG_CHANNEL_CLOSE without having sent or received SSH_MSG_CHANNEL_EOF." This sounds to me that even if an EOF has been sent, a CLOSE may be sent on the same channel.

Thus, a ``shutdown `write`` followed by `close` will sent first a SSH_MSG_CHANNEL_EOF and then a SSH_MSG_CHANNEL_CLOSE.

Basically, what @reynir said.

Reply:
But in contrast to what @reynir said, the changes in this PR (so far) are only affecting the client side -- the server side does not use the read/write or flow functionality (as far as I can tell).

Reply:
I pushed a commit that does not call `FLOW.shutdown` for the above reasons; but that may have been rushed. I will look into what the semantics are...

Reply:
> the application doesn't write anymore

It's not: _the application can not write anymore_? I mean, if I follow `write_flow` and we fallback to the `Error `Closed`, we set the state to `Writer_closed`, use `write_flow`/`writev_flow` then in the `Writer_closed` seems wrong for me where we will probably fallback _again_ to the `Error `Closed` when we do the `FLOW.write`. We probably should justify our last `write_flow`/`writev_flow` and say that we deliberately ignore the result (which is what we do with `>|= fun _`).

Reply:
Indeed, we mix up two things here:
- the underlying TCP state (where if a `write` returns `` `Closed``, this will never recover)
- the application calling ``shutdown `write`` (which means "the application will never write again")

Both result in t.state being ``Writer_closed``

In a similar vain, if `read` returns `` `Eof``, we move to `` `Reader_closed`` -- same if we ``shutdown `read`` -- here as well, the former means that any subsequent `read` will return `` `Eof`` (or error), while in the latter scenario we can still read (e.g. CHANNEL_EOF / CHANNEL_CLOSE).

Now, the question is: should we distinguish these cases? Is it worth it? Won't TCP always return `` `Closed `` when writing to a closed connection? I think the states are already in this PR exploding (from `` `Active | `Eof | `Error `` to `` `Active | `Reader_closed | `Writer_closed | `Closed | `Error ``). And surely we can also invalidate the underlying TCP socket (by making the t.flow an option).

Reply:
Probably just a comment to say in comment that the `writev_flow` in the `Writed_closed` can fail and we ignore and are aware about that :+1:, it's not totally obvious for a newcomer.

Reply:
I pushed 611d6b4, hope that makes it clear?